pkg/handlers/save_test.go

```diff
@@ -0,0 +1,240 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := context.WithValue(r.Context(), middleware.OptlyClientKey, suite.oc)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+func (suite *SaveTestSuite) SetupTest() {
+	testClient := optimizelytest.NewClient()
+	userProfileService := &userprofileservices.InMemoryUserProfileService{
+		ProfilesMap: make(map[string]decision.UserProfile),
+	}
+	optlyClient := &optimizely.OptlyClient{
+		OptimizelyClient:   testClient.OptimizelyClient,
+		ConfigManager:      nil,
+		ForcedVariations:   testClient.ForcedVariations,
+		UserProfileService: userProfileService,
+	}
+
+	mux := chi.NewMux()
+	mux.With(suite.ClientCtx).Post("/save", Save)
+
+	body := saveBody{
+		UPSResponseOut: UPSResponseOut{
+			UserID: "testUser",
+		},
+	}
+	payload, err := json.Marshal(body)
+	suite.NoError(err)
+
+	suite.body = payload
+	suite.mux = mux
+	suite.tc = testClient
+	suite.oc = optlyClient
+}
+
+func (suite *SaveTestSuite) TestNilPayload() {
+	req := httptest.NewRequest("POST", "/save", nil)
+	rec := httptest.NewRecorder()
+	suite.mux.ServeHTTP(rec, req)
+
+	suite.assertError(rec, ErrEmptyUserID.Error(), http.StatusBadRequest)
+}
+
+func (suite *SaveTestSuite) TestInvalidPayload() {
+	req := httptest.NewRequest("POST", "/save", bytes.NewBuffer([]byte("ABC€")))
+	rec := httptest.NewRecorder()
+	suite.mux.ServeHTTP(rec, req)
+	suite.assertError(rec, "error parsing request body", http.StatusBadRequest)
+}
+
+func (suite *SaveTestSuite) TestNoUserProfileService() {
+	req := httptest.NewRequest("POST", "/save", bytes.NewBuffer(suite.body))
+	rec := httptest.NewRecorder()
+	suite.oc.UserProfileService = nil
+	suite.mux.ServeHTTP(rec, req)
+
+	suite.assertError(rec, ErrNoUPS.Error(), http.StatusInternalServerError)
+}
+
+func (suite *SaveTestSuite) TestSaveEmptyUserID() {
+	body := saveBody{
+		UPSResponseOut: UPSResponseOut{
+			UserID: "",
+			ExperimentBucketMap: map[string]interface{}{
+				"1": map[string]interface{}{"variation_id": "2"},
+				"2": map[string]interface{}{"variation_id": "2"},
+				"3": map[string]interface{}{"variation_id": "3"},
+			},
+		},
+	}
+	payload, err := json.Marshal(body)
+	suite.NoError(err)
+
+	req := httptest.NewRequest("POST", "/save", bytes.NewBuffer(payload))
+	rec := httptest.NewRecorder()
+	suite.mux.ServeHTTP(rec, req)
+
+	suite.assertError(rec, ErrEmptyUserID.Error(), http.StatusBadRequest)
+}
+
+func (suite *SaveTestSuite) TestSaveEmptyBucketMap() {
+	req := httptest.NewRequest("POST", "/save", bytes.NewBuffer(suite.body))
+	rec := httptest.NewRecorder()
+	suite.mux.ServeHTTP(rec, req)
+
+	suite.Equal(http.StatusOK, rec.Code)
+
+	body := saveBody{
+		UPSResponseOut: UPSResponseOut{
+			UserID: "testUser",
+		},
+	}
+
+	expected := convertToUserProfile(body)
+	actual := suite.oc.UserProfileService.Lookup("testUser")
+	suite.Equal(expected, actual)
+}
+
+func (suite *SaveTestSuite) TestSave() {
```

Comment:
`save` only checks for `userId` to be non-empty, in which case it directly saves the userprofile against the `userId` in the map. There's no other validation to it. This can be verified from the following test: https://github.com/optimizely/agent/pull/326/files#diff-495a600fac0bb299d45a5c0b0c193c4d3ce65eff2f38123d927f4ced74a2d0a9R195